### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by samjones1001**

### CODE REVIEW

File: src/hooks/useHttpRequest/index.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
+
+export interface responseObject {
+  response?: AxiosResponse,
+  complete: boolean,
+  error: boolean
+}
+
+const useHttpRequest = () => {
+  const [res, setRes] = useState<responseObject>({
```
[the remark this answers is not in the document]
Comment (MollieS) [reply src/hooks/useHttpRequest/index.tsx:11]: Ah of course.  I think using a custom hook should be fine, I guess I'm just wondering what exactly is the problem this is solving?  If it's react-y to use the custom hook, go ahead.  Is it convention to use state within here though? I feel like something that is just making an HTTP call shouldn't have to hold onto state, surely it just wants to pass the info back to the component who needs it?  I may well be missing something though so 🤷‍♀

File: src/components/SignUpForm/index.tsx [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+import FormInput, { FormInputProps } from '../FormInput';
+import FormDropdown, { FormDropdownProps, OptionProps } from "../FormDropdown";
+
+import useHttpRequest from '../../hooks/useHttpRequest';
+import userHeadersHelpers, { userHeaders } from "../../helpers/userHeadersHelper";
+import authenticationService from "../../services/authenticationService";
+
+import './index.scss';
+
+const SignUpForm: FC = ({}) => {
```
[the remark this answers is not in the document]
Comment (samjones1001) [reply src/components/SignUpForm/index.tsx:16]: I've tried to attack this a bit and have removed some of the excess:

- I've created an AuthForm component which is responsible for rendering all of the various elements which make up the form itself. Both the SignInForm and SignUpForm components now delegate rendering to this component.
- I've extracted the logic for signing in and authenticating a user out to the AuthenticationService, and removed some repeated logic from both of the form components.
- I've simplified some of the logic around managing the stage of the SignUpForm.
- I've extracted some of the constants holding information about the form inputs to render out to a constants.tsx file.

File: src/components/SignUpForm/index.tsx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    if (res.complete && !res.error) {
+      signIn(res.response)
+    }
+  });
+
+  const signIn = ((res?: AxiosResponse) => {
+    if (res) {
+      const headerObj: userHeaders = userHeadersHelpers.responseHeadersToObject(res.headers); 
+      authenticationService.login(headerObj);
+      setIsAuthenticated(authenticationService.isAuthenticated());
```
[the remark this answers is not in the document]
Comment (samjones1001) [reply src/components/SignUpForm/index.tsx:37]: This has now been removed from both of the form components.

File: src/hooks/useHttpRequest/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (samjones1001) [reply src/hooks/useHttpRequest/index.tsx:11]: I've looked at this a bit more and have found a way of making the hook work without making use of setState.

I'm inclined to agree that making the hook stateful feels like the wrong thing to do but explanations I've found online all seem to follow the pattern I originally had here so I'm not sure. Take a look at the change I made and if this seems better to you then I'll stick with it as the result is the same.

**Pull request merged**